[FEATURE] spring oauth2 client와 jwt를 활용한 카카오 소셜로그인 구현

SOULIKK opened the pull request:
spring oauth2 client와 jwt를 활용한 카카오 소셜로그인 구현

JWT, oauth2-client, redis 의존성을 추가했습니다
설정 파일에 oauth2와 토큰 키에 대한 설정을 추가했습니다
redis를 활용한 refresh 토큰의 저장 및 관리 는 캐시 전략에 대해 논의 후 개선해나가겠습니다

SOULIKK committed:
spring oauth2 client와 jwt를 활용한 카카오 소셜로그인 구현

redis를 활용한 refresh 토큰의 저장 및 관리 는 캐시 전략에 대해 논의 후 개선해나가겠습니다

OldRabbit736 commented:
민성님 잘 구동 되던가요?

GitHub Actions의 integration test가 실패해서,
로컬로 받아서 실행 시도해 봤는데, 다음과 같은 에러가 발생하네요,
org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'securityConfig':

내용을 살펴보니 oauth2의 client-id 가 설정이 안되어 있다는 것 같습니다.

application.yml의 client-id, client-secret에 값이 필요한 거 같습니다.

SteadyFlow736 reviewed (dismissed):
민성님 어플리케이션 잘 구동 되던가요?

GitHub Actions의 integration test가 실패해서,
로컬로 받아서 실행 시도해 봤는데, 다음과 같은 에러가 발생하네요,
org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'securityConfig':

내용을 살펴보니 oauth2의 client_id 값이 없다는 에러 같습니다.
application.yml의 client_id 값이 없는 것과 관련 있는 것 같습니다.

SOULIKK commented:
> 



> 민성님 어플리케이션 잘 구동 되던가요?
> 
> GitHub Actions의 integration test가 실패해서, 로컬로 받아서 실행 시도해 봤는데, 다음과 같은 에러가 발생하네요, org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'securityConfig':
> 
> 내용을 살펴보니 oauth2의 client_id 값이 없다는 에러 같습니다. application.yml의 client_id 값이 없는 것과 관련 있는 것 같습니다.



> 민성님 어플리케이션 잘 구동 되던가요?
> 
> GitHub Actions의 integration test가 실패해서, 로컬로 받아서 실행 시도해 봤는데, 다음과 같은 에러가 발생하네요, org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'securityConfig':
> 
> 내용을 살펴보니 oauth2의 client_id 값이 없다는 에러 같습니다. application.yml의 client_id 값이 없는 것과 관련 있는 것 같습니다.

SOULIKK commented:
public repo에 키값을 노출시키지 않으려고 일단 공란으로 두었습니다
yaml 파일을 분리하고 노출되지 않게 관리할 수 있도록 수정하겠습니다

SteadyFlow736 commented:
> public repo에 키값을 노출시키지 않으려고 일단 공란으로 두었습니다 yaml 파일을 분리하고 노출되지 않게 관리할 수 있도록 수정하겠습니다

아하, 그렇군요. 키 값 노출을 생각지 못했네요. 알겠습니다. 저도 일단 로컬에서 카카오 로그인 서비스 이용해서 따로 테스트 해 보겠습니다.

SteadyFlow736 commented on app/src/main/java/com/foodlab/foodReservation/auth/config/OAuth2SuccessHandler.java:
```diff
@@ -0,0 +1,45 @@
+        response.addHeader("accessToken", token.getAccessToken());
+        response.addHeader("refreshToken", token.getRefreshToken());
+        response.setContentType("application/json;charset=UTF-8");
+
```
토큰을 헤더와 바디 둘 다 넣어주고 있는데, 이유가 무엇인지 궁금합니다!
또한 setContentType 메서드가 두 번 실행되고 있습니다.

SteadyFlow736 commented on app/src/main/java/com/foodlab/foodReservation/auth/config/SecurityConfig.java:
```diff
@@ -0,0 +1,38 @@
+                .antMatchers("/token/**", "/login/oauth2/code/kakao").permitAll()
+                .anyRequest().authenticated()
+                .and()
+                .oauth2Login().loginPage("/token/expired")
```
OAuth2 로그인 페이지가 설정되어 있는데, 추후 OAuth2 로그인 페이지도 넣으실 생각이신지요?

SOULIKK replied:
처음에 넣으려다 제외했습니다 loginPage부분을 삭제하겠습니다

SteadyFlow736 commented on app/src/main/java/com/foodlab/foodReservation/auth/config/UserInfo.java:
```diff
@@ -0,0 +1,57 @@
+
+@Getter
+@NoArgsConstructor
+public class UserInfo implements UserDetails {
```
이 클래스는 다른 클래스와 연관은 없는 건가요? (Customer 클래스 등) 어떤 용도인지 궁금합니다!

SOULIKK replied:
@Authenticationprincipal로 로그인한 유저의 정보를 참조하기 위한 UserDetails를 구현클래스입니다!

SteadyFlow736 commented on app/src/main/java/com/foodlab/foodReservation/customer/entity/Customer.java:
```diff
@@ -28,5 +28,8 @@ public class Customer {
 
+    public Customer(String email) {
+        this.email = email;
+    }
```
IDE 에서는 Customer 클래스에 다음의 에러를 표시하고 있습니다.
"Class 'Customer' should have [public, protected] no-arg constructor"
아무래도 public Customer(String email) 컨스터럭터가 추가되면서 기본 컨스트럭터가 없어져 버려서 생긴 문제인 것 같습니다.
@Entity 어노테이션은 기본 컨스트럭터를 필요로 하기 때문입니다.
@NoArgsConstructor(access = AccessLevel.PROTECTED) 을 클래스에 붙여주면 문제가 해결될 것 같습니다.

SOULIKK replied:
넵 추가하겠습니다!

SteadyFlow736 reviewed (commented):
몇 가지 궁금한 사항을 코멘트 하였습니다!

SOULIKK committed:
리뷰 반영 및 리팩토링

회원정보를 세션에 추가하는 로직을 제거했습니다
@WebMvcTest가 OAuth2Client 빈을 스캔하지 못해 발생한 에러(NoSuchBeanDefinitionException)에 대해 컴포넌트 필터 설정을 추가해 처리했습니다
spring-security-test 의존성을 추가, mock user를 활용해 integration test를 수행 가능하도록 수정했습니다
회원을 조회, 등록(DB에 없는 경우) 로직을 OAuth2SuccessHandler에서 수행하도록 수정했습니다 (토큰 발급 전 회원 데이터를 조회, 저장할 수 있도록)
토큰 claims에 id, email을 같이 넣어서 리턴하도록 수정했습니다
토큰 검증시 prefix (Bearer ) 체크를 추가했습니다
Customer 엔티티에 기본생성자 설정(NoArgsConstructor)을 추가했습니다

SOULIKK committed:
Merge branch 'main' of https://github.com/SOULIKK/food-reservation into feature/auth

SOULIKK committed:
Merge branch 'f-lab-edu:main' into feature/auth

SOULIKK committed:
Merge branch 'main' of https://github.com/SOULIKK/food-reservation into feature/auth

# Conflicts:
#	app/build.gradle
#	app/src/main/resources/application.yml

SOULIKK committed:
auth 설정 추가

SOULIKK committed:
Merge branch 'feature/auth' of https://github.com/SOULIKK/food-reservation into feature/auth

SOULIKK committed:
auth 설정파일 경로 수정

SteadyFlow736 commented:
수고 하셨습니다!

SteadyFlow736 commented:
민성님, 애플리케이션 구동 과정에서 다음과 같은 에러가 발생하는 것 같습니다.

# error1
"Could not resolve placeholder 'token-secret-key' in value "${token-secret-key}""
아무래도 JwtProvider의 secretKey 필드가 값을 받아오지 못하는 것 같습니다.
설정 파일에 추가가 필요합니다! (token-refresh-key 도 마찬가지로 추가가 필요할듯하네요)

# error2
token-secret-key, token-refresh-key 를 dummy 값으로 지정하고 시도해본 결과,
이번엔 client-secret 값을 바인딩하지 못했다고 오류가 발생합니다.
"Failed to bind properties under 'spring.security.oauth2.client.registration.kakao.client-secret' to java.lang.String"
혹시나 싶어 jasypt의 decrypt.sh로 확인해본 결과 현재 들어있는 ENC 값이 뭔가 이상한 듯 합니다.
client-id ENC 값은 decrypt.sh 와 저희가 정해두었던 비밀번호에 의해 잘 decrypt가 되었지만
client-secret ENC 값은 decrypt.sh로 decrypt 시도 시 오류가 발생합니다.
아무래도 encrypt 과정에서 뭔가 잘못된 것 같습니다.
다시 한번 ENC 값을 확인하고 업데이트가 필요할 거 같습니다.
